apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Span.java

```diff
@@ -292,13 +303,132 @@ public void beforeEnd(long epochMicros) {
 
     @Override
     protected void afterEnd() {
-        this.tracer.endSpan(this);
+        if (transaction != null && transaction.isSpanCompressionEnabled() && parent != null) {
+            Span buffered = parent.bufferedSpan.get();
+            if (!isCompressionEligible()) {
+                if (buffered != null) {
+                    if (parent.bufferedSpan.compareAndSet(buffered, null)) {
+                        this.tracer.endSpan(buffered);
+                    }
+                }
+                this.tracer.endSpan(this);
+                return;
+            }
+            if (buffered == null) {
+                if (!parent.bufferedSpan.compareAndSet(null, this)) {
+                    // the failed update would ideally lead to a compression attempt with the new buffer,
+                    // but we're dropping the compression attempt to keep things simple
+                    this.tracer.endSpan(this);
+                }
+                return;
+            }
+            if (!buffered.tryToCompress(this)) {
+                if (parent.bufferedSpan.compareAndSet(buffered, this)) {
+                    this.tracer.endSpan(buffered);
+                } else {
+                    // the failed update would ideally lead to a compression attempt with the new buffer,
+                    // but we're dropping the compression attempt to keep things simple
+                    this.tracer.endSpan(this);
+                }
+            } else if (isSampled()) {
+                Transaction transaction = getTransaction();
+                if (transaction != null) {
+                    transaction.getSpanCount().getDropped().incrementAndGet();
+                }
+                decrementReferences();
+            }
+        } else {
+            this.tracer.endSpan(this);
+        }
+    }
+
+    private boolean isCompressionEligible() {
+        return isExit() && isDiscardable() && (outcomeNotSet() || getOutcome() == Outcome.SUCCESS);
+    }
+
+    private boolean tryToCompress(Span sibling) {
+        boolean canBeCompressed = isComposite() ? tryToCompressComposite(sibling) : tryToCompressRegular(sibling);
+        if (!canBeCompressed) {
+            return false;
+        }
+
+        long newDuration = sibling.getTimestamp() + sibling.duration.get() - getTimestamp();
+        do {
+            long currentDuration = duration.get();
+            if (newDuration <= currentDuration) {
+                break;
+            }
+            if (duration.compareAndSet(currentDuration, newDuration)) {
+                break;
+            }
+        } while (true);
+
+        composite.increaseCount();
+        composite.increaseSum(sibling.duration.get());
+
+        return true;
+    }
+
+    private boolean tryToCompressRegular(Span sibling) {
+        if (!isSameKind(sibling)) {
+            return false;
+        }
+
+        long currentDuration = duration.get();
```

Comment:
move down to after the immediately following clause as it's not needed until then